### PR TITLE
Improve applicable batched gemms

### PR DESCRIPTION
**Pull request by pfultz2**
This can help remove the need for contiguous for many cases when using BS 1 for bert.

### CODE REVIEW

**Comment by causten** [comment]
I tested this branch and it does indeed boost performance for BS1

File: src/targets/gpu/gemm_impl.cpp [hunk 1]
```diff
@@ -27,6 +28,22 @@ rocblas_datatype get_type(shape::type_t type)
     MIGRAPHX_THROW("ROCBLAS_GEMM: data type not supported!");
 }
 
+void blas_shape(const shape& s)
+{
+    if(s.lens().size() < 2)
```
Comment (umangyadav) [inline src/targets/gpu/gemm_impl.cpp:33]: It is checking for size() < 2 and size() < 3, returning both times, Should it only check size() < 3 ?

File: src/targets/gpu/gemm_impl.cpp [hunk 2]
```diff
@@ -27,6 +28,22 @@ rocblas_datatype get_type(shape::type_t type)
...
+        return;
+    if(std::none_of(s.strides().end() - 2, s.strides().end(), [&](auto i) { return i == 1; }))
```
Comment (umangyadav) [inline src/targets/gpu/gemm_impl.cpp:35]: Should be lengths instead of strides to be collapsible ? I think both works.

File: src/reduce_dims.cpp [hunk 3]
```diff
@@ -37,17 +35,33 @@ bool reduce_dim(std::vector<shape>& shapes, std::size_t n)
     return true;
 }
 
+void reduce_dim1(std::vector<shape>& shapes)
```
Comment (umangyadav) [inline src/reduce_dims.cpp:38]: This looks similar to `squeeze` op on last axis.

Comment (pfultz2) [reply src/reduce_dims.cpp:38]: Yea, since the other `reduce_dim` doesn't work on the last dimension.

File: src/targets/gpu/gemm_impl.cpp (hunk 1 above)
Comment (pfultz2) [reply src/targets/gpu/gemm_impl.cpp:33]: No it should return yet. If the size is at least two we need to check the memory layout still,

File: src/targets/gpu/gemm_impl.cpp (hunk 2 above)
Comment (pfultz2) [reply src/targets/gpu/gemm_impl.cpp:35]: This isn't checking the batch dimensions. Blas needs one of the strides of the matrix dimension(ie either the row or column) to be `1`, so this checking for that.

**Review by umangyadav — APPROVED** [review]

**Review by kahmed10 — APPROVED** [review]

**Pull request merged**